### DISCUSSION TITLE
VipsOperation: class "gifsave_buffer" not found

### DISCUSSION THREAD

**Discussion by Jacherr** [libvips/libvips#3960] [Q&A] [answered]
Hi,

I'm attempting to use libvips from C, and it's been going great so far! Thank you for such an easy-to-use API. 🙏

One small issue I've hit is attempting to save GIFs, whereupon I get the error `VipsOperation: class "gifsave_buffer" not found` from the error buffer. `cgif` should be installed, and in fact I updated libvips to be sure and indeed it says `GIF save with cgif                : YES` when performing `meson setup build --prefix /usr`. Could it still be down to a bad installation of `cgif` or something else? Any help appreciated.

**Comment by jcupitt** [comment]
Hi @Jacherr,

That's strange. Maybe you have two versions of libvips installed, with an older version being picked up by your program at runtime?

You can use `vips -l` and `--version` at the command-line to list all operations and show the version you have, eg.:

```
john@banana ~/GIT/nip4/build (main) $ vips -l | grep gifsave
        VipsForeignSaveCgif (gifsave_base), save as gif (.gif), priority=0, rgba-only
          VipsForeignSaveCgifFile (gifsave), save as gif (.gif), priority=0, rgba-only
          VipsForeignSaveCgifBuffer (gifsave_buffer), save as gif (.gif), priority=0, rgba-only
          VipsForeignSaveCgifTarget (gifsave_target), save as gif (.gif), priority=0, rgba-only
john@banana ~/GIT/nip4/build (main) $ vips --version
vips-8.16.0
john@banana ~/GIT/nip4/build (main) $ which vips
/home/john/vips/bin/vips
```

Verify which libvips you program is compiling against with:

```
john@banana ~/GIT/nip4/build (main) $ pkg-config vips --cflags
-I/home/john/vips/include -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -I/usr/include/libmount -I/usr/include/blkid -I/usr/include/webp -I/usr/include/pango-1.0 -I/usr/include/harfbuzz -I/usr/include/freetype2 -I/usr/include/libpng16 -I/usr/include/fribidi -I/usr/include/cairo -I/usr/include/pixman-1 -I/usr/include/x86_64-linux-gnu -I/usr/include/librsvg-2.0 -I/usr/include/gdk-pixbuf-2.0 -I/usr/include/hdf5/serial -I/usr/include/OpenEXR -pthread -I/usr/include/Imath -I/usr/include/openjpeg-2.5 -DHWY_SHARED_DEFINE -I/usr/include/nifti 
john@banana ~/GIT/nip4/build (main) $ pkg-config vips --modversion
8.16.0
```

And finally see which libvips you are picking up at runtime with `ldd` and by calling `vips_version_string()`.

**Reply by Jacherr** [reply]
Ah yep, it does appear I might have multiple installs unfortunately. Any idea what the best way would be to prune out old ones? Unfortunately not very familiar with installation processes for libraries like this.

**Reply by Jacherr** [reply]
After trying several times to completely remove all old installations, it's still somehow getting an old version 8.15.1 from somewhere despite me removing files from `/usr/bin/vips, /usr/include/vips, /usr/share/man/man1/vips.1`, and also removing all old pkg-config files. This seems to be affecting both command-line vips and the C library.

I have absolutely no idea where it's getting this from, because if I remove all of these files, `vips` is no longer a found command and my program won't build because it cant find vips. However, if I then install 8.15.2, with `cgif`, it then starts reporting `8.15.1` again with no gif saving support. Really bizarre.

**Comment by jcupitt** [comment, marked as answer]
It sounds like you installed with `--prefix /usr`, which is usually a mistake, though it depends a bit on your platform.

The files in `/usr` are managed for you by your package manager. If you start adding and removing files as well, it'll get tangled up and chaos will follow. Package installs will start to fail, updates will not complete successfully, mysterious crashes will begin.

I'd try to clean up with:

1. Use your package manager to remove any libvipses which it thinks exist.
2. Install libvips to a prefix in your home area, perhaps with `meson setup build --prefix ~/vips`, then have a look though the `~/vips` directory and see what files it created. 
3. Look for those same files in `/usr/lib` etc. and make sure you remove them all.
4. Finally, install libvips again to a safe prefix, see below.

There's a chance you'll have damaged your install too badly to fix and you'll have to reinstall the OS :( Hopefully you'll be lucky.

When installing from source, there are two usual prefixes to use:

1. A system-wide prefix, `/usr/local`. This is the meson default, and the official prefix for local software installs. You could also use `/usr/local/vips` if you want all the vips files in one place. After installing here, you may or may not need to update your system in various ways, it depends on your platform. 
2. A user prefix, `~/vips`. This will install for just you. I usually pick this for personal use, since I'm not interested in a system-wide install, and having all the files in one place makes it very simple to uninstall with `rm -rf ~/vips`. 

When installing packages from source, there are three things you need to adjust or check in your environment:

1. You need the `vips` executable on your `PATH`. 
2. You need `vips.pc` on your `PKG_CONFIG_PATH`.
3. You need `libvips.so` on your library path. For a system-wide install, you need to read up on the run-time linker and `/etc/ld.so.conf`. For a user install, you need to add to `LD_CONFIG_PATH`.

I have this in my `.bashrc`, for example:

```bash
export VIPSHOME=/home/john/vips
export PATH="$VIPSHOME/bin:$PATH"
export PKG_CONFIG_PATH="$VIPSHOME/lib/pkgconfig:$PKG_CONFIG_PATH"
export LD_LIBRARY_PATH="$VIPSHOME/lib"
```

**Reply by Jacherr** [reply]
Ah, this all makes sense. Might just be easier to reinstall the OS - thankfully, its just Windows Subsystem for Linux, so not much is going to be lost. Probably would have been good to figure out an appropriate prefix first - not sure what I was trying to do!

Many thanks for the help, I'll give all of this a shot now.

**Reply by Jacherr** [reply]
After a bit of fiddling around in the `/usr` folders, I think I've managed to get rid of all the old files. vips 8.15.2 is reported along with the `gifsave` API! Once again, thank you for the help as I bumble around with C library installation.

**Reply by jcupitt** [reply]
WSL has a pretty recent ubuntu, so you probably don't need to build libvips from source. Just enter `sudo apt install libvips-dev` and it'll do everything for you automatically.
